internal/hook/wait_exec_hook_handler.go

```diff
@@ -0,0 +1,260 @@
+				maxWait = hook.Hook.WaitTimeout.Duration
+			}
+		}
+	}
+	if maxWait > 0 {
+		ctx, cancel = context.WithTimeout(ctx, maxWait)
+	}
+	waitStart := time.Now()
+
+	var errors []error
+
+	handler := func(newObj interface{}) {
+		newPod, ok := newObj.(*v1.Pod)
+		if !ok {
+			return
+		}
+
+		if newPod.Status.Phase == v1.PodSucceeded || newPod.Status.Phase == v1.PodFailed {
+			err := fmt.Errorf("Pod %s/%s entered phase %s before some post-restore exec hooks ran", newPod.Namespace, newPod.Name, newPod.Status.Phase)
+			log.Warning(err)
+			cancel()
+			return
+		}
+
+		for containerName, hooks := range byContainer {
+			if !isContainerRunning(newPod, containerName) {
+				continue
+			}
+			podMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(newPod)
+			if err != nil {
+				log.WithError(err).Errorf("error unstructuring pod %s/%s", newPod.Namespace, newPod.Name)
+				cancel()
+				return
+			}
+
+			// Sequentially run all hooks for the ready container. The container's hooks are not
+			// removed from the byContainer map until all have completed so that if one fails
+			// remaining unexecuted hooks can be handled by the outer function.
+			for i, hook := range hooks {
+				// This indicates to the outer function not to handle this hook as unexecuted in
+				// case of terminating before deleting this container's slice of hooks from the
+				// byContainer map.
+				byContainer[containerName][i].executed = true
+
+				hookLog := log.WithFields(
+					logrus.Fields{
+						"hookSource": hook.Source,
+						"hookType":   "exec",
+						"hookPhase":  "post",
+					},
+				)
+				// Check the individual hook's wait timeout is not expired
+				if hook.Hook.WaitTimeout.Duration != 0 && time.Since(waitStart) > hook.Hook.WaitTimeout.Duration {
+					err := fmt.Errorf("Hook %s in container %s in pod %s/%s expired before executing", hook.Name, hook.Hook.Container, pod.Namespace, pod.Name)
+					hookLog.Error(err)
+					if hook.Hook.OnError == velerov1api.HookErrorModeFail {
+						errors = append(errors, err)
+						cancel()
+						return
+					}
+				}
+				eh := &velerov1api.ExecHook{
+					Container: hook.Hook.Container,
+					Command:   hook.Hook.Command,
+					OnError:   hook.Hook.OnError,
+					Timeout:   hook.Hook.ExecTimeout,
+				}
+				if err := e.PodCommandExecutor.ExecutePodCommand(hookLog, podMap, pod.Namespace, pod.Name, hook.Name, eh); err != nil {
+					hookLog.WithError(err).Error("Error executing hook")
+					if hook.Hook.OnError == velerov1api.HookErrorModeFail {
+						errors = append(errors, err)
+						cancel()
+						return
+					}
+				}
+			}
+			delete(byContainer, containerName)
+		}
+		if len(byContainer) == 0 {
+			cancel()
+		}
+	}
+
+	selector := fields.OneTermEqualSelector("metadata.name", pod.Name)
+	lw := e.ListWatchFactory.NewListWatch(pod.Namespace, selector)
+
+	_, controller := cache.NewInformer(lw, pod, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: handler,
+		UpdateFunc: func(_, newObj interface{}) {
```

Comment:
This means that `handler` will be called once when the pod is created and again when the pod is scheduled and is running.
Can we simplify this by registering only `UpdateFunc` and calling the `handler` only when the pod is `Running`?
We should also keep the `DeleteFunc`

Reply:
In my tests the pod always got passed to `AddFunc` the first time and then `UpdateFunc` for any changes. I think it means added to the local cache, not to the API. So if we don't watch Add and the pod doesn't change it could be missed entirely.

Reply:
When a Pod is created, it starts with a `Pending` phase. This is when `AddFunc` handler will be called.
Once the Pod is scheduled and transitions to a `Running` phase, the `UpdateFunc` handler will be called.

It is possible that the pod may have transitioned into the `Running` phase by the time the handler for `AddFunc` can run. When this happens the hook is already run.
When the pod transitions into `Running` it will also queue the update event which will invoke the `UpdateFunc` handler which will run the hook again and this may be undesirable. This approach will guarantee "at least once" semantics for the hook execution.

I am thinking we should register a handler for only `AddFunc`, and not the `UpdateFunc` and wait for the pod to start `Running` and then execute the rook. This will give us "exactly-once" semantics for hook execution.
I can't remember, from the design doc, what semantics we said that these hooks will follow.

Reply:
Are there other cases in the update event that we should be handling? what if there are updates that we don't really care about? Will that too trigger the execution of the restore hooks?

Reply:
I've added a comment above the handler about how it ensures exactly-once semantics if invoked on both AddFunc and UpdateFunc. Updates to a pod are ignored unless it's either 1. a container starts running or 2. the pod enters a phase that means no more containers in it will ever start running.

Reply:
Thank you for clarifying and adding the comments.
What about the scenario in which:
1. `AddFunc` handler is triggered and is currently running the hook.
2. While the `AddFunc` handler is running, if there is any update to the pod, that will trigger the `UpdateFunc` handler which will also try to run the hook.
This will cause a race between two executions of the hook.

I don't know if the Informer interface does resource de-dup and ensure only one of the handlers is running. I know the controller runtime does perform the de-dup.

Reply:
Looking at https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L389
Seems like only one of the `OnUpdate`, `OnAdd`, or `OnDelete` functions can be called at a time and that this function should complete before the next `Delta` is handled.
So there shouldn't be a race in the scenario I described.

Reply:
@ashish-amarnath Is this resolved, then?

Reply:
yes, this is resolved.